<h1> Quiz 1 : Pemahaman Tentang Model Evaluasi</h1>

Jawab pertanyaan di bawah ini dengan bahasa kalian masing2?

1. Apa perbedaan antara data latih, data validasi, dan data test?
2. Bagaimana cara kita menilai performa suatu model?
3. Apa itu Confusion Matriks? Jelaskan secara lengkap!
4. Apa itu Classification Report dari sklearn?

<h2> 1. Apa perbedaan antara data latih, data validasi, dan data test?</h2>

* Data latih merupakan data yang digunakan untuk melatih model agar model dapat mengerjakan suatu permasalahan tertentu, biasa dijatahi 75% dari total data
* Data Validasi merupakan data yang digunakan untuk melakukan validasi sebelum pada akhirnya dilakukan testing model menggunakan data testing, biasa dijatahi 10% dari data latih
* Data testing merupakan data yang digunakan melakukan testing pada model, biasa dijatahi 25% dari total data

<h2> 2. Bagaimana cara kita menilai performa suatu model? </h2>

* Untuk klasifikasi, kita bisa melihat dari tingkat akurasi, F1 score dan juga dari confusion matrix model itu sendiri
* sedangkan untuk regresi biasa menggunakan tingkat akurasi dan juga RMSE

<h2> 3. Apa itu confusion Matriks? Jelaskan secara lengkap! </h2>

* confusion matrix merupakan suatu matrix yg menampilkan performa model dalam melakukan prediksi, terdapat 4 titik yaitu (True Positif, False Positif, False Negativ, True Negatif)

* True Positif: model menebak bahwa labelnya adalah positif dan benar positif

* False Positif: model menebak bahwa labelnya adalah negatif namun yg benar adalah positif

* False Negatif: model menebak bahwa labelnya adalah positif namun yg benar adalah negatif

* True Negatif: model menebak bahwa labelnya adalah negatif dan benar negatif

<h2> 4. Apa itu classification report dari sklearn? </h2>
* classification report merupakan module metrics dari sklearn yang digunakan untuk menampilkan beberapa parameter pengukuran model klasifikasi seperti recall, precision, f1 score dan support

<h1>Quiz 2 : Applikasi Model Evaluasi</h1>

Kali ini kita akan menggunakan data untuk memprediksi kelangsungan hidup pasien yang telah mengalami operasi payu dara. Dengan informasi yang dimiliki terkait pasien, kita akan membuat model untuk memprediksi apakah pasien akan bertahan hidup dalam waktu lebih dari 5 tahun atau tidak.

Lebih Lengkapnya kalian bisa membaca informasi tentang dataset di link berikut : https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.names

Buat model Klasifikasi (Model/Algoritma Bebas) untuk memprediksi status pasien. dengan ketentuan sebagai berikut :

1. Bagi kedua data ini menjadi data training dan data test dengan test_size=0.25.
3. Pelajar tentang metrics roc_auc_score kemudian buatlah model dan evaluasi dengan menggunakan teknik cross-validation dengan scoring 'roc_auc'. baca https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html untuk menggunakan metric roc_auc saat cross-validation.
3. Berapa score rata2 dari model dengan teknik cross-validation tersebut?
4. Prediksi data test dengan model yang telah kalian buat!
5. Bagaimana hasil confusion matrix dari hasil prediksi tersebut?
6. Bagaimana classification report dari hasil prediksi tersebut?
5. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status positive?
6. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status negatif?

In [1]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [2]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [3]:
df['survival_status'].value_counts()

1    225
2     81
Name: survival_status, dtype: int64

In [4]:
dfsatu  = df[df['survival_status'] == 1]
dfdua   = df[df['survival_status'] == 2]

dfsatu  = dfsatu[:len(dfdua)]
df      = pd.concat([dfsatu, dfdua])

In [5]:
df['survival_status'].value_counts()

2    81
1    81
Name: survival_status, dtype: int64

In [6]:
y = df['survival_status']
X = df.drop('survival_status', axis=1)

In [7]:
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.preprocessing import StandardScaler

In [8]:
for i in X.columns:
    X[i] = StandardScaler().fit_transform(X[[i]])

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, stratify=y)

lr = LogisticRegression()
lr.fit(X_train, y_train)
print(f"Score rata2 CV 10 fold adalah\t: {cross_val_score(lr, X, y, cv=10, scoring='roc_auc').mean()}")

Score rata2 CV 10 fold adalah	: 0.9902777777777778


In [10]:
y_pred  = lr.predict(X_test)

print(f'Hasil prediksi\t: {y_pred}')

Hasil prediksi	: [1 1 1 2 2 1 2 2 2 1 1 2 2 2 2 2 1 1 2 2 1 2 2 1 2 1 1 1 1 2 1 1 1 1 2 1 1
 2 1 2 1]


In [11]:
print('Hasil dari Confusion Matrix')
print(confusion_matrix(y_test, y_pred))

print('\n\nHasil dari Classification report')
print(classification_report(y_test, y_pred))

Hasil dari Confusion Matrix
[[17  3]
 [ 5 16]]


Hasil dari Classification report
              precision    recall  f1-score   support

           1       0.77      0.85      0.81        20
           2       0.84      0.76      0.80        21

    accuracy                           0.80        41
   macro avg       0.81      0.81      0.80        41
weighted avg       0.81      0.80      0.80        41



* Label 1 memiliki f1 score sebesar 89%
* Label 2 memiliki f1 score sebesar 85%